[core] use internal `/run` for pid and pickle files

degemer opened the pull request:
- use `/run` for run files (supervisor, pid and pickle), move more functions to `utils/`
- start `dogstatd` and `dd-agent` only when needed
- remove legacy option clean

See commits descriptions for more details.

It should affect:
- https://github.com/DataDog/dd-agent/pull/1504 (needs to add the `mkdir $dd_agent/run` step)
- https://github.com/DataDog/dd-agent/pull/1361 (needs to rebase afterwards)

And of course https://github.com/DataDog/dd-agent-omnibus, to add a internal `/run` directory (Windows is not concerned)

degemer commented:
This PR should also allow us to run `supervisor` as `dd-agent`, and thus fix https://github.com/DataDog/dd-agent/issues/1348

LeoCavaille commented:
Makes sense to me.
However, I'll maybe still allow `Platform` (and `PidFile` although less likely) to be imported from where they were before. Nothing says they're part of a private API, and people might use those in custom checks to switch behaviors on different OSs, you can leave a compat import, clearly state that it is deprecated and add a `FIXME` for 6.x.

degemer commented:
Sure, I added them back. I wanted to deprecate them, but it's a pain since all `Platform` methods are class methods, so I just added a comment instead.

remh commented:
@degemer can you rebase please ?

remh commented on agent.py:
```diff
@@ -77,6 +77,7 @@ def _handle_sigusr1(self, signum, frame):
         self._do_restart()
 
+    @classmethod
     def info(self, verbose=None):
```
Tiny tiny nitpick but if you're switching to a class method, it should be 
```
cls
```
instead of 
```
self
```

remh commented on daemon.py:
```diff
@@ -227,7 +227,7 @@ def run(self):
 
-
+    @classmethod
     def info(self):
```
same here

remh commented on packaging/centos/datadog-agent.init:
```diff
@@ -31,7 +31,7 @@ FORWARDERPATH="/opt/datadog-agent/agent/ddagent.py"
 SUPERVISORCTL_PATH="/opt/datadog-agent/bin/supervisorctl"
 SUPERVISOR_CONF="/etc/dd-agent/supervisor.conf"
-SUPERVISOR_SOCK="/var/tmp/datadog-supervisor.sock"
+SUPERVISOR_SOCK="/opt/datadog-agent/run/datadog-supervisor.sock"
```
It shouldn't be an issue, but we'll need to triple check that it doesn't cause any trouble while upgrading.

remh commented on utils/tailfile.py:
```diff
@@ -1,29 +1,16 @@
-import base64
+import binascii
 import os
+from stat import *
```
Let's just import what we need ?

remh commented:
Great work overall! Just a bunch of tiny nitpicks and it's good to merge.

degemer commented:
Thanks Rémi! Updated and rebased.
